### PR TITLE
Fix the bug of predicate will change after select

### PR DESCRIPTION
**Pull request by amymeng5277**

### CODE REVIEW

File: src/foam/dao/MDAO.java [hunk 1]
```diff
@@ -69,19 +69,29 @@ public FObject find_(X x, Object o) {
 
   public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
     SelectPlan plan;
-    if ( predicate != null ) predicate = predicate.partialEval();
-    if ( predicate instanceof Or ) {
-      int length = ( (Or) predicate ).getArgs().length;
+    Predicate copyPredicate = null;
+    if ( predicate != null ) {
+      copyPredicate = (Predicate) ( (FObject) predicate ).deepClone();
```
Comment (kgrgreer) [inline src/foam/dao/MDAO.java]: Why do you need to clone?

Comment (amymeng5277) [reply src/foam/dao/MDAO.java]: Because the predicate will change after we deal with it or did partialEval()

Comment (kgrgreer) [reply src/foam/dao/MDAO.java]: partialEval should create a new predicate, not mutate itself

Comment (amymeng5277) [reply src/foam/dao/MDAO.java]: So you mean we don't need use predicate = predicate.partialEval(), we use it directly? Actually we return a new predicate.

**Pull request merged**